agent/consul/filter.go

```diff
@@ -22,9 +22,13 @@ func (d *dirEntFilter) Move(dst, src, span int) {
 
 // FilterDirEnt is used to filter a list of directory entries
 // by applying an ACL policy
-func FilterDirEnt(acl acl.ACL, ent structs.DirEntries) structs.DirEntries {
+func FilterDirEnt(acl acl.ACL, ent structs.DirEntries) (structs.DirEntries, error) {
 	df := dirEntFilter{acl: acl, ent: ent}
-	return ent[:FilterEntries(&df)]
+	filtered := ent[:FilterEntries(&df)]
+	if len(filtered) == 0 {
+		return filtered, errPermissionDenied
```

Comment:
For all of these cases I'd return `nil, errPermissionDenied`, just to make it stand out that the filtered results aren't usable. It'll work the same way, but this makes it clear.